edxadapt/edxadapt.py

```diff
@@ -0,0 +1,137 @@
+import logging
+import pkg_resources
+
+from django.template import Template, Context
+from HTMLParser import HTMLParser
+from xblock.core import XBlock
+from xblock.fields import Scope, Boolean, Dict, List, String
+from xblock.fragment import Fragment
+from xblockutils.studio_editable import StudioEditableXBlockMixin
+
+from student.models import anonymous_id_for_user, User
+
+
+_ = lambda text: text
+log = logging.getLogger(__name__)
+html_parser = HTMLParser()
+
+
+class EdxAdaptXBlock(StudioEditableXBlockMixin, XBlock):
+    """
+    This XBlock automatically registers students in the EdxAdapt instance
+    using parameters configured by the teaching staff: params, skills, edx_adapt_api_url
+    """
+
+    has_author_view = True
+
+    display_name = String(
+        default='Adaptive Learning',
+        scope=Scope.content,
+    )
+
+    params = Dict(
+        default={'pg': 0.25, 'ps': 0.25, 'pi': 0.1, 'pt': 0.5, 'threshold': 0.99},
+        display_name=_('BKT params'),
+        help=_('Barameters for Bayesian Knowledge Tracing (BKT) model'),
+        scope=Scope.content
+    )
+
+    skills = List(
+        default=[
+            'center', 'shape', 'spread', 'x axis', 'y axis', 'h to d',
+            'd to h', 'histogram', 'None'
+        ],
+        display_name=_('Course skills list'),
+        scope=Scope.content,
+        help=_(
+            'List of skills of this course. Each problem addresses certain skill. '
+            'Special skill "None" is used for those problems which belong to none.'
+        )
+    )
+    edx_adapt_api_url = String(
+        default='',
+        display_name=_('Edx Adapt API base URL'),
+        scope=Scope.content,
+        help=_('Edx Adapt API base URL, e.g. https://edx-adapt.example.com:443/api/v1')
+    )
+
+    student_is_registered = Boolean(
+        default=False,
+        scope=Scope.preferences
+    )
+
+    editable_fields = ('params', 'skills', 'edx_adapt_api_url')
+
+    def resource_string(self, path):
+        """Handy helper for getting resources from our kit."""
+        data = pkg_resources.resource_string(__name__, path)
+        return data.decode("utf8")
+
+    def ugettext(self, text):
+        """
+        For backward compatibility with older version of XBlock module released before Mar 22, 2016
+
+        See https://github.com/edx/XBlock/blame/ab636a82bc23c3f65a618e438637d518b209790f/xblock/core.py#L204
+        """
+        return text
+
+    def get_anonymous_student_id(self):
+        """
+        Returns course non-specific anonymous student id compatible with
+        anonymous id used by Capa Problems.
+        """
+
+        return anonymous_id_for_user(
+            User.objects.get(
+                id=self.runtime.user_id), None
```

Comment:
I'm not happy about this part too. And yes, runtime does provide `anonymous_student_id`.

Unfortunately it's a "course specific" anonymous id. Whereas we need "course non specific" anonymous id, because only this type of anonymous id is available in a capa-problems context, when students solve adaptive problems.

Here's comparison:
```python
# Provided by XBlock runtime
ipdb> self.runtime.anonymous_student_id
'58cfd06b8e049a0e599c9eea166f8857'

# Course non-specific anonymous id
ipdb> anonymous_id_for_user(User.objects.get(id=self.runtime.user_id), None)
'66c6d171816ef81dc88859f0e87ae732'

# Course specific anonymous id
ipdb> anonymous_id_for_user(User.objects.get(id=self.runtime.user_id), self.course_id)
'58cfd06b8e049a0e599c9eea166f8857'
```